File: packages/arb-avm-cpp/avm/src/checkpoint/transaction.cpp
/*
 * Copyright 2019, Offchain Labs, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <avm/checkpoint/transaction.hpp>

#include <avm/checkpoint/checkpointresult.hpp>
#include <avm/checkpoint/checkpointutils.hpp>

#include <rocksdb/utilities/transaction_db.h>

Transaction::Transaction(rocksdb::Transaction* transaction_) {
    transaction = std::unique_ptr<rocksdb::Transaction>(transaction_);
}

Transaction::~Transaction() {
    assert(transaction->GetState() == rocksdb::Transaction::COMMITED ||
           rocksdb::Transaction::ROLLEDBACK);
}

rocksdb::Status Transaction::commit() {
    return transaction->Commit();
}

rocksdb::Status Transaction::rollBack() {
    return transaction->Rollback();
}

SaveResults Transaction::incrementReference(
    const std::vector<unsigned char>& hash_key) {
    auto results = getValue(hash_key);

    if (results.status.ok()) {
        auto updated_count = results.reference_count + 1;
        return saveValueWithRefCount(updated_count, hash_key,
                                     results.stored_value);
    } else {
        return SaveResults{0, results.status, hash_key};
    }
}

SaveResults Transaction::saveValue(const std::vector<unsigned char>& hash_key,
                                   const std::vector<unsigned char>& value) {
    auto results = getValue(hash_key);
    int ref_count;

    if (results.status.ok()) {
        assert(results.stored_value == value);
        ref_count = results.reference_count + 1;
    } else {
        ref_count = 1;
    }
    return saveValueWithRefCount(ref_count, hash_key, value);
};

DeleteResults Transaction::deleteValue(
    const std::vector<unsigned char>& hash_key) {
    auto results = getValue(hash_key);

    if (results.status.ok()) {
        if (results.reference_count < 2) {
            auto delete_status = deleteKeyValuePair(hash_key);
            return DeleteResults{0, delete_status};

        } else {
            auto updated_ref_count = results.reference_count - 1;
            auto update_result = saveValueWithRefCount(
                updated_ref_count, hash_key, results.stored_value);
            return DeleteResults{updated_ref_count, update_result.status};
        }
    } else {
        return DeleteResults{0, results.status};
    }
}

GetResults Transaction::getValue(
    const std::vector<unsigned char>& hash_key) const {
    auto read_options = rocksdb::ReadOptions();
    std::string return_value;
    std::string key_str(hash_key.begin(), hash_key.end());
    auto get_status = transaction->Get(read_options, key_str, &return_value);

    if (get_status.ok()) {
        auto tuple = checkpoint::storage::parseCountAndValue(return_value);
        auto stored_val = std::get<1>(tuple);
        auto ref_count = std::get<0>(tuple);

        return GetResults{ref_count, get_status, stored_val};
    } else {
        auto unsuccessful = rocksdb::Status().NotFound();
        return GetResults{0, unsuccessful, std::vector<unsigned char>()};
    }
}

// private
// ----------------------------------------------------------------------
SaveResults Transaction::saveValueWithRefCount(
    uint32_t updated_ref_count,
    const std::vector<unsigned char>& hash_key,
    const std::vector<unsigned char>& value) {
    auto updated_entry =
        checkpoint::storage::serializeCountAndValue(updated_ref_count, value);

    auto status = saveKeyValuePair(hash_key, updated_entry);

    if (status.ok()) {
        return SaveResults{updated_ref_count, status, hash_key};
    } else {
        return SaveResults{0, status, hash_key};
    }
}

rocksdb::Status Transaction::saveKeyValuePair(
    const std::vector<unsigned char>& key,
    const std::vector<unsigned char>& value) {
    std::string value_str(value.begin(), value.end());
    std::string key_str(key.begin(), key.end());
    return transaction->Put(key_str, value_str);
}

rocksdb::Status Transaction::deleteKeyValuePair(
    const std::vector<unsigned char>& key) {
    std::string key_str(key.begin(), key.end());
    return transaction->Delete(key_str);
}

File: packages/arb-validator/message/transaction.go
/*
* Copyright 2020, Offchain Labs, Inc.
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
 */

package message

import (
	"bytes"
	"errors"
	"math/big"

	"github.com/offchainlabs/arbitrum/packages/arb-validator/evm"

	"github.com/offchainlabs/arbitrum/packages/arb-util/common"
	"github.com/offchainlabs/arbitrum/packages/arb-util/hashing"
	"github.com/offchainlabs/arbitrum/packages/arb-util/value"
)

type Transaction struct {
	Chain       common.Address
	To          common.Address
	From        common.Address
	SequenceNum *big.Int
	Value       *big.Int
	Data        []byte
}

func (m Transaction) Equals(o Transaction) bool {
	return m.Chain == o.Chain &&
		m.To == o.To &&
		m.From == o.From &&
		m.SequenceNum.Cmp(o.SequenceNum) == 0 &&
		m.Value.Cmp(o.Value) == 0 &&
		bytes.Equal(m.Data, o.Data)
}

func (m Transaction) Type() MessageType {
	return TransactionType
}

func (m Transaction) AsValue() value.Value {
	val1, _ := value.NewTupleFromSlice([]value.Value{
		addressToIntValue(m.To),
		value.NewIntValue(m.Value),
	})
	val2, _ := value.NewTupleFromSlice([]value.Value{
		value.NewIntValue(big.NewInt(int64(m.Type()))),
		addressToIntValue(m.From),
		val1,
	})
	return val2
}

func (m Transaction) ReceiptHash() common.Hash {
	return hashing.SoliditySHA3(
		hashing.Uint8(uint8(m.Type())),
		hashing.Address(m.To),
		hashing.Address(m.From),
		hashing.Uint256(m.Value),
		hashing.Uint256(m.SequenceNum),
		m.Data,
	)
}

type DeliveredTransaction struct {
	Transaction
	BlockNum *common.TimeBlocks
}

func (m DeliveredTransaction) Equals(other DeliveredMessage) bool {
	o, ok := other.(DeliveredTransaction)
	if !ok {
		return false
	}
	return m.Transaction.Equals(o.Transaction) &&
		m.BlockNum.Cmp(o.BlockNum) == 0
}

func (m DeliveredTransaction) DeliveredHeight() *common.TimeBlocks {
	return m.BlockNum
}

func (m DeliveredTransaction) CommitmentHash() common.Hash {
	return hashing.SoliditySHA3(
		hashing.Uint8(uint8(m.Type())),
		hashing.Address(m.To),
		hashing.Address(m.From),
		hashing.Uint256(m.Value),
		hashing.Uint256(m.SequenceNum),
		hashing.Uint256(m.BlockNum.AsInt()),
		m.Data,
	)
}

func (m DeliveredTransaction) CheckpointValue() value.Value {
	val, _ := value.NewTupleFromSlice([]value.Value{
		addressToIntValue(m.Chain),
		addressToIntValue(m.To),
		addressToIntValue(m.From),
		value.NewIntValue(m.SequenceNum),
		value.NewIntValue(m.Value),
		evm.BytesToByteStack(m.Data),
		value.NewIntValue(m.BlockNum.AsInt()),
	})
	return val
}

func UnmarshalTransaction(v value.Value) (DeliveredTransaction, error) {
	tup, ok := v.(value.TupleValue)
	if !ok || tup.Len() != 7 {
		return DeliveredTransaction{}, errors.New("tx val must be 7-tuple")
	}
	chain, _ := tup.GetByInt64(0)
	chainInt, ok := chain.(value.IntValue)
	if !ok {
		return DeliveredTransaction{}, errors.New("chain must be int")
	}
	to, _ := tup.GetByInt64(1)
	toInt, ok := to.(value.IntValue)
	if !ok {
		return DeliveredTransaction{}, errors.New("to must be int")
	}
	from, _ := tup.GetByInt64(2)
	fromInt, ok := from.(value.IntValue)
	if !ok {
		return DeliveredTransaction{}, errors.New("from must be int")
	}
	seq, _ := tup.GetByInt64(3)
	seqInt, ok := seq.(value.IntValue)
	if !ok {
		return DeliveredTransaction{}, errors.New("seq must be int")
	}
	val, _ := tup.GetByInt64(4)
	valInt, ok := val.(value.IntValue)
	if !ok {
		return DeliveredTransaction{}, errors.New("chain must be int")
	}
	data, _ := tup.GetByInt64(5)
	dataBytes, err := evm.ByteStackToHex(data)
	if err != nil {
		return DeliveredTransaction{}, err
	}
	blockNum, _ := tup.GetByInt64(6)
	blockNumInt, ok := blockNum.(value.IntValue)
	if !ok {
		return DeliveredTransaction{}, errors.New("blockNum must be int")
	}

	return DeliveredTransaction{
		Transaction: Transaction{
			Chain:       intValueToAddress(chainInt),
			To:          intValueToAddress(toInt),
			From:        intValueToAddress(fromInt),
			SequenceNum: seqInt.BigInt(),
			Value:       valInt.BigInt(),
			Data:        dataBytes,
		},
		BlockNum: common.NewTimeBlocks(blockNumInt.BigInt()),
	}, nil
}
